File: agents/skills/claude-code/canary-katana/scripts/alarm.py
#!/usr/bin/env python3
"""alarm -- fire only when a deletion removes the last coverage of a hot symbol.

The premise that keeps this skill from becoming nag fatigue: most test deletions
are legitimate, so katana is silent by default and records everything. It alarms
in exactly one situation -- the deleted test was the *last* test covering a
symbol that ``critical-areas.json`` marks as high-risk. When that file is
missing or malformed the alarm degrades to recording-only and says so out loud;
a gate that manufactures failures on missing data is one people mute, and a
muted gate is worse than no gate.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from enum import Enum
from pathlib import Path

import diffscan

DEGRADED_NOTICE = "critical-area data unavailable, recording only, not alarming"

# risk_score at or above this makes a name-matched last-coverage loss CRITICAL;
# below it the loss is still real but ranked HIGH.
_CRITICAL_RISK = 0.7

# Directory names too generic to imply a coverage relationship on their own.
_GENERIC_DIRS = {
    "src", "lib", "app", "apps", "packages", "pkg",
    "tests", "test", "__tests__", "e2e", "spec", "dist", "build",
}

_CODE_SUFFIXES = (".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs", ".py")


class Fidelity(Enum):
    NAME_MATCHED = "name-matched"
    HEURISTIC = "heuristic"

    @property
    def rank(self) -> int:
        return 0 if self is Fidelity.NAME_MATCHED else 1


class Severity(Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"

    @property
    def sort_key(self) -> int:
        return {"critical": 0, "high": 1, "medium": 2}[self.value]


@dataclass
class CriticalAreas:
    available: bool
    areas: list
    reason: str


@dataclass
class Finding:
    kind: str
    test: str
    file: str
    area: str
    fidelity: Fidelity
    severity: Severity
    evidence: str

    def to_dict(self) -> dict:
        return {
            "kind": self.kind,
            "test": self.test,
            "file": self.file,
            "area": self.area,
            "fidelity": self.fidelity.value,
            "severity": self.severity.value,
            "evidence": self.evidence,
        }


def load_critical_areas(path) -> CriticalAreas:
    """Load ``critical-areas.json``; unavailable (not raising) on any problem."""
    if path is None:
        return CriticalAreas(False, [], "critical-area file not provided")
    path = Path(path)
    if not path.exists():
        return CriticalAreas(False, [], f"critical-area file not found: {path}")
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError) as exc:
        return CriticalAreas(False, [], f"critical-area file malformed: {exc}")
    areas = data.get("areas", []) if isinstance(data, dict) else []
    return CriticalAreas(True, list(areas), "")


def _norm(text: str) -> str:
    return re.sub(r"[^a-z0-9]", "", text.lower())


def area_symbols(area_path: str) -> set:
    """Symbols an area path exposes: its basename minus code suffix, and parts.

    ``src/loyalty/points.service.ts`` -> {"points.service", "points", "service"}.
    """
    base = area_path.replace("\\", "/").split("/")[-1]
    for suffix in _CODE_SUFFIXES:
        if base.endswith(suffix):
            base = base[: -len(suffix)]
            break
    symbols = {base}
    symbols.update(part for part in base.split(".") if part)
    return symbols


def _area_norm_symbols(area_path: str) -> set:
    return {s for s in (_norm(x) for x in area_symbols(area_path)) if len(s) >= 4}


def _dirs(path: str) -> list:
    parts = [p for p in path.replace("\\", "/").split("/") if p]
    return parts[:-1]


def _significant_dirs(path: str) -> set:
    return {d for d in _dirs(path) if d not in _GENERIC_DIRS}


def _name_covers(test_name: str, norm_symbols: set) -> bool:
    normalized = _norm(test_name)
    return any(sym in normalized for sym in norm_symbols)


def _repo_test_files(repo: Path):
    for path in sorted(repo.rglob("*")):
        if not path.is_file():
            continue
        rel = path.relative_to(repo).as_posix()
        if diffscan.is_test_file(rel):
            yield rel, path


_PY_TEST_DEF = re.compile(r"^\s*(?:async\s+)?def\s+(test\w*)\s*\(", re.MULTILINE)
_JS_TEST_CALL = re.compile(
    r"\b(?:describe|context|it|test)(?:\.\w+)?\s*\(\s*(['\"`])(.*?)\1"
)


def _test_names(text: str) -> list:
    names = [m.group(1) for m in _PY_TEST_DEF.finditer(text)]
    names += [m.group(2) for m in _JS_TEST_CALL.finditer(text)]
    return names


def _name_coverage_remains(repo: Path, norm_symbols: set) -> bool:
    for _rel, path in _repo_test_files(repo):
        try:
            text = path.read_text(encoding="utf-8")
        except OSError:
            continue
        if any(_name_covers(name, norm_symbols) for name in _test_names(text)):
            return True
    return False


def _dir_coverage_remains(repo: Path, area_dirs: set) -> bool:
    for rel, path in _repo_test_files(repo):
        if not (set(_dirs(rel)) & area_dirs):
            continue
        try:
            text = path.read_text(encoding="utf-8")
        except OSError:
            continue
        if _test_names(text):
            return True
    return False


def build_findings(deletions, areas: CriticalAreas, repo) -> list:
    """Return last-coverage-removed findings; empty when data is unavailable."""
    if not areas.available:
        return []  # silent by default: never alarm on degraded data
    repo = Path(repo)
    findings: list[Finding] = []

    for deletion in deletions:
        best: Finding = None
        for area in areas.areas:
            area_path = area.get("path", "")
            risk = float(area.get("risk_score", 0.0) or 0.0)
            norm_symbols = _area_norm_symbols(area_path)

            if norm_symbols and _name_covers(deletion.name, norm_symbols):
                if _name_coverage_remains(repo, norm_symbols):
                    continue
                fidelity = Fidelity.NAME_MATCHED
                severity = (
                    Severity.CRITICAL if risk >= _CRITICAL_RISK else Severity.HIGH
                )
            else:
                area_dirs = _significant_dirs(area_path)
                if not (area_dirs & set(_dirs(deletion.file))):
                    continue
                if _dir_coverage_remains(repo, area_dirs):
                    continue
                fidelity = Fidelity.HEURISTIC
                severity = Severity.MEDIUM

            candidate = Finding(
                kind="last-coverage-removed",
                test=deletion.name,
                file=deletion.file,
                area=area_path,
                fidelity=fidelity,
                severity=severity,
                evidence=(
                    f"{deletion.name} was the last test covering {area_path}"
                ),
            )
            if best is None or (fidelity.rank, severity.sort_key) < (
                best.fidelity.rank,
                best.severity.sort_key,
            ):
                best = candidate

        if best is not None:
            findings.append(best)

    findings.sort(key=lambda f: (f.severity.sort_key, f.file, f.test))
    return findings

File: agents/skills/claude-code/canary-katana/scripts/alarm.mjs
// alarm -- fire only when a deletion removes the last coverage of a hot symbol.
//
// Most test deletions are legitimate, so katana is silent by default and
// records everything. It alarms in exactly one situation: the deleted test was
// the *last* test covering a symbol `critical-areas.json` marks high-risk. When
// that file is missing or malformed the alarm degrades to recording-only and
// says so; a gate that manufactures failures on missing data gets muted, and a
// muted gate is worse than no gate.

import fs from 'node:fs';

import { isTestFile } from './diffscan.mjs';

export const DEGRADED_NOTICE =
  'critical-area data unavailable, recording only, not alarming';

// risk_score at or above this makes a name-matched last-coverage loss CRITICAL;
// below it the loss is still real but ranked HIGH.
const CRITICAL_RISK = 0.7;

// Directory names too generic to imply a coverage relationship on their own.
const GENERIC_DIRS = new Set([
  'src',
  'lib',
  'app',
  'apps',
  'packages',
  'pkg',
  'tests',
  'test',
  '__tests__',
  'e2e',
  'spec',
  'dist',
  'build',
]);

const CODE_SUFFIXES = ['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs', '.py'];

export const Fidelity = {
  NAME_MATCHED: { value: 'name-matched', rank: 0 },
  HEURISTIC: { value: 'heuristic', rank: 1 },
};

export const Severity = {
  CRITICAL: { value: 'critical', sortKey: 0 },
  HIGH: { value: 'high', sortKey: 1 },
  MEDIUM: { value: 'medium', sortKey: 2 },
};

/**
 * @typedef {{available: boolean, areas: Array<Record<string, any>>, reason: string}} CriticalAreas
 */

/** JSON-contract shape of a finding. */
export function findingToDict(f) {
  return {
    kind: f.kind,
    test: f.test,
    file: f.file,
    area: f.area,
    fidelity: f.fidelity.value,
    severity: f.severity.value,
    evidence: f.evidence,
  };
}

/**
 * Load critical-areas.json; unavailable (not throwing) on any problem.
 * @returns {CriticalAreas}
 */
export function loadCriticalAreas(filePath) {
  if (filePath === null || filePath === undefined) {
    return {
      available: false,
      areas: [],
      reason: 'critical-area file not provided',
    };
  }
  if (!fs.existsSync(filePath)) {
    return {
      available: false,
      areas: [],
      reason: `critical-area file not found: ${filePath}`,
    };
  }
  let data;
  try {
    data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch (exc) {
    return {
      available: false,
      areas: [],
      reason: `critical-area file malformed: ${exc.message}`,
    };
  }
  const areas =
    data &&
    typeof data === 'object' &&
    !Array.isArray(data) &&
    Array.isArray(data.areas)
      ? data.areas
      : [];
  return { available: true, areas: [...areas], reason: '' };
}

const norm = (text) => text.toLowerCase().replace(/[^a-z0-9]/g, '');

/**
 * Symbols an area path exposes: its basename minus code suffix, plus parts.
 * `src/loyalty/points.service.ts` -> {points.service, points, service}.
 */
export function areaSymbols(areaPath) {
  let base = areaPath.replace(/\\/g, '/').split('/').pop();
  for (const suffix of CODE_SUFFIXES) {
    if (base.endsWith(suffix)) {
      base = base.slice(0, -suffix.length);
      break;
    }
  }
  const symbols = new Set([base]);
  for (const part of base.split('.')) if (part) symbols.add(part);
  return symbols;
}

const areaNormSymbols = (areaPath) => {
  const out = new Set();
  for (const s of areaSymbols(areaPath)) {
    const n = norm(s);
    if (n.length >= 4) out.add(n);
  }
  return out;
};

const dirsOf = (p) => {
  const parts = p.replace(/\\/g, '/').split('/').filter(Boolean);
  return parts.slice(0, -1);
};

const significantDirs = (p) =>
  new Set(dirsOf(p).filter((d) => !GENERIC_DIRS.has(d)));

const nameCovers = (testName, normSymbols) => {
  const normalized = norm(testName);
  return [...normSymbols].some((sym) => normalized.includes(sym));
};

// Heavy/ignored directories never worth walking for test files (#395).
const SKIP_DIRS = new Set([
  '.git',
  'node_modules',
  '__pycache__',
  '.venv',
  'venv',
  'dist',
  'build',
  '.mypy_cache',
  '.pytest_cache',
  '.tox',
  'coverage',
  '.next',
  '.turbo',
]);

/**
 * Enumerate repo test files, pruning heavy dirs. Deterministic (sorted).
 * @returns {[string, string][]} [relPosixPath, absolutePath] pairs.
 */
export function repoTestFiles(repo) {
  const results = [];
  const walk = (absDir, relDir) => {
    let entries;
    try {
      entries = fs.readdirSync(absDir, { withFileTypes: true });
    } catch {
      return;
    }
    const dirs = [];
    const files = [];
    for (const e of entries) {
      if (e.isDirectory()) {
        if (!SKIP_DIRS.has(e.name)) dirs.push(e.name);
      } else if (e.isFile()) {
        files.push(e.name);
      }
    }
    for (const name of files.sort()) {
      const rel = relDir ? `${relDir}/${name}` : name;
      if (isTestFile(rel)) results.push([rel, `${absDir}/${name}`]);
    }
    for (const name of dirs.sort()) {
      walk(`${absDir}/${name}`, relDir ? `${relDir}/${name}` : name);
    }
  };
  walk(String(repo), '');
  return results;
}

const PY_TEST_DEF = /^\s*(?:async\s+)?def\s+(test\w*)\s*\(/gm;
const JS_TEST_CALL =
  /\b(?:describe|context|it|test)(?:\.\w+)?\s*\(\s*(['"`])(.*?)\1/g;

function testNames(text) {
  const names = [];
  for (const m of text.matchAll(PY_TEST_DEF)) names.push(m[1]);
  for (const m of text.matchAll(JS_TEST_CALL)) names.push(m[2]);
  return names;
}

const readTextSafe = (p) => {
  try {
    return fs.readFileSync(p, 'utf8');
  } catch {
    return null;
  }
};

function nameCoverageRemains(repo, normSymbols) {
  for (const [, abs] of repoTestFiles(repo)) {
    const text = readTextSafe(abs);
    if (text === null) continue;
    if (testNames(text).some((name) => nameCovers(name, normSymbols)))
      return true;
  }
  return false;
}

function dirCoverageRemains(repo, areaDirs) {
  for (const [rel, abs] of repoTestFiles(repo)) {
    if (!dirsOf(rel).some((d) => areaDirs.has(d))) continue;
    const text = readTextSafe(abs);
    if (text === null) continue;
    if (testNames(text).length) return true;
  }
  return false;
}

/** Return last-coverage-removed findings; empty when data is unavailable. */
export function buildFindings(deletions, areas, repo) {
  if (!areas.available) return []; // silent by default: never alarm on degraded data
  const findings = [];

  for (const deletion of deletions) {
    let best = null;
    for (const area of areas.areas) {
      const areaPath = area.path || '';
      const risk = Number.parseFloat(area.risk_score) || 0.0;
      const normSymbols = areaNormSymbols(areaPath);

      let fidelity;
      let severity;
      if (normSymbols.size && nameCovers(deletion.name, normSymbols)) {
        if (nameCoverageRemains(repo, normSymbols)) continue;
        fidelity = Fidelity.NAME_MATCHED;
        severity = risk >= CRITICAL_RISK ? Severity.CRITICAL : Severity.HIGH;
      } else {
        const areaDirs = significantDirs(areaPath);
        const delDirs = new Set(dirsOf(deletion.file));
        if (![...areaDirs].some((d) => delDirs.has(d))) continue;
        if (dirCoverageRemains(repo, areaDirs)) continue;
        fidelity = Fidelity.HEURISTIC;
        severity = Severity.MEDIUM;
      }

      const candidate = {
        kind: 'last-coverage-removed',
        test: deletion.name,
        file: deletion.file,
        area: areaPath,
        fidelity,
        severity,
        evidence: `${deletion.name} was the last test covering ${areaPath}`,
      };
      // Keep the best candidate per deletion: lower (fidelity.rank, sortKey)
      // wins, element-wise (name-matched outranks heuristic; then severity).
      if (best === null) {
        best = candidate;
      } else {
        const better =
          fidelity.rank !== best.fidelity.rank
            ? fidelity.rank < best.fidelity.rank
            : severity.sortKey < best.severity.sortKey;
        if (better) best = candidate;
      }
    }
    if (best !== null) findings.push(best);
  }

  findings.sort(
    (a, b) =>
      a.severity.sortKey - b.severity.sortKey ||
      a.file.localeCompare(b.file) ||
      a.test.localeCompare(b.test),
  );
  return findings;
}
